[BEAM-12000] Update deprecated collections to use collections.abc in python sdk

hmc-cs-mdrissi opened the pull request:
Update all usages of collections.Sequence/collections.Iterable/etc to use collections.abc as old name is deprecated and will be removed in 3.10. collections.abc was added in 3.3 so all python versions beam supports should be compatible with this change.

@tvalentyn This fix was motivated by seeing the deprecation warning and this ticket, https://issues.apache.org/jira/browse/BEAM-12000?jql=text%20~%20%223.9%22. Doesn't solve the main work for the ticket, but covers one easy change.

tvalentyn reviewed (commented):
Thanks a lot for this fix.

Should we import collecitons.abc as well? I am seeing tests are passing, so not sure if all the affected codepaths were triggered, but seeing this locally:

```
:~$ python3
Python 3.9.2 (default, Feb 28 2021, 17:03:44) 
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import collections
>>> isinstance(object, collections.abc.Iterable)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.9/collections/__init__.py", line 68, in __getattr__
    raise AttributeError(f'module {__name__!r} has no attribute {name!r}')
AttributeError: module 'collections' has no attribute 'abc'
>>> import collections.abc
>>> isinstance(object, collections.abc.Iterable)
False
```

pabloem commented:
what's the status for this PR?

hmc-cs-mdrissi commented:
I've updated the import to cover collections.abc. I'm unsure why the tests are fine as some tests definitely do pass by those lines based on adding breakpoints, but better to be safe. Most likely reason why tests are fine is that some other import indirectly adds collections.abc. It's a common library to import so most of the time you will get somewhere.

PR status was mainly I needed to get dev environment set up. Back when I made the pr I made the edits without setting up an environment. Now that I've got beam locally set up, I added a simple unit test that should hopefully fix coverage failure. Coverage failure was only thing blocking the pr earlier. For coverage I noticed overload lines were counting as untyped when they should be skipped so added overload to exclude_lines

hmc-cs-mdrissi commented on sdks/python/apache_beam/utils/proto_utils.py:
```diff
@@ -40,13 +40,13 @@
 def pack_Any(msg):
   # type: (message.Message) -> any_pb2.Any
-  pass
+  ...
```
pass vs ... are similar but not same. pass has an inferred return type of None for some type checkers. Pyright will give you an error for using pass here. ... is typical way to mark overloads as stubs.

hmc-cs-mdrissi replied:
Are test files even type checked? I tried running mypy on this file and got an error. Also discovered mypy.ini for the repo has an invalid value giving this error message,

`mypy.ini: [mypy]: follow_imports: invalid choice 'true' (choose from 'normal', 'silent', 'skip', 'error')`

The mypy error that prevents it from type checking this file is,

`apache_beam/portability/api/metrics_pb2.pyi:117: error: invalid syntax  [syntax]` Looks like root cause of that error is a comment using type: notation for doc string and confusing it with standard type comment.

tvalentyn replied:
Interesting, I think generated _pb2.py files should be excluded from the type checks.

tvalentyn replied:
Didn't know about this usage of Ellipsis. Found: https://mypy.readthedocs.io/en/stable/stubs.html and https://github.com/python/typing/issues/109 where this is disussed.

hmc-cs-mdrissi replied:
The pb2.py is not type checked, but pb2.pyi is needed for type inference. Looks like proto file has a comment that confuses mypy. I can make a tweak to the comment to make it fine for mypy.

hmc-cs-mdrissi commented on .pre-commit-config.yaml:
```diff
@@ -42,3 +42,8 @@ repos:
         args: ["--rcfile=sdks/python/.pylintrc"]
         files: ^sdks/python/apache_beam/
         exclude: *exclude
+  - repo: https://github.com/pycqa/isort
```
Added as CI runs isort but pre-commit doesn't. Should have same behavior as ci. The setup.cfg tweaks I did were to match run_pylint.sh in CI.

hmc-cs-mdrissi commented:
I'm confused by test failure on ubuntu-latest 3.7, but 3.7 works on other platforms + other ubuntus work. None of the changes I made look platform related. Can that check be re-run/potentially flaky?

tvalentyn commented:
> I'm confused by test failure on ubuntu-latest 3.7, but 3.7 works on other platforms + other ubuntus work. None of the changes I made look platform related. Can that check be re-run/potentially flaky?

It's a flake, it's being fixed in BEAM-12794.

> Why is there both .coveragerc and setup.cfg with coverage options? Which one is used by CI? Can there only be one?

I don't know, you could ask on dev@ or @udim might know.

tvalentyn commented on sdks/python/apache_beam/utils/proto_utils_test.py:
```diff
@@ -0,0 +1,25 @@
+    orig_msg = unpack_Any(packed_msg, timestamp_pb2.Timestamp)
+    none_msg = unpack_Any(packed_msg, None)
+    assert proto_timestamp == orig_msg
+    assert none_msg is None
```
is this a left over given that you have a separate test_none_pack?

hmc-cs-mdrissi replied:
test_None_pack is to check packing None. unpack_Any does not allow None for the msg, but allows None for the msg type. So this is intended. I wanted to test both None msg pack and None class unpack.

tvalentyn commented on sdks/python/setup.cfg:
```diff
@@ -53,7 +56,11 @@ exclude_lines =
 output = target/site/cobertura/coverage.xml
 
 [isort]
+line_length = 120
```
where does this number come from?

hmc-cs-mdrissi replied:
https://github.com/apache/beam/blob/master/sdks/python/scripts/run_pylint.sh#L105 it's chosen to match isort usage in CI.

tvalentyn replied:
Came across https://issues.apache.org/jira/browse/BEAM-3745 with touches on isort & lint. perhaps we can close it already or after your changes.

tvalentyn commented on sdks/python/.coveragerc:
```diff
@@ -16,4 +16,23 @@
+
+  # Don't complain about missing debug-only code:
+  def __repr__
+  if self\.debug
```
how about: `if self\.debug_logging_enabled` to avoid pattern-matching with `if self.debug_options` which occurs in the codebase?

hmc-cs-mdrissi replied:
Sure. The reason I did if self\.debug was to make .coveragerc consistent with setup.cfg coverage option. For the infra pr I can make them both self\.debug_logging_enabled

tvalentyn commented on sdks/python/apache_beam/utils/proto_utils_test.py:
```diff
@@ -0,0 +1,25 @@
+import unittest
```
RAT check wants you to add a license here, see other files.

hmc-cs-mdrissi replied:
Thank you will fix.

tvalentyn commented:
Thanks, I left some comments. It would be better to split infra changes in this PR into self-contained changes, either as separate PRs or separate commits appropriately squashed. If you choose to have multiple commits in the same PR, please don't squash the commits before review on this PR finalizes.

hmc-cs-mdrissi commented:
I'll split the pr into two prs. One for infra changes and one for unit test + abc.collections change. Unit test is only needed with the abc.collections change for coverage check to be satisfied.

pabloem commented:
hi there! any updates on this PR? : )

tvalentyn commented:
Closing in favor of https://github.com/apache/beam/pull/15850 @hmc-cs-mdrissi feel free to rebase this PR to keep the infrastructure changes that you made and reopen it. Thank you!
